crates/ingress-client/src/session.rs

```diff
@@ -0,0 +1,501 @@
+                        break SessionState::Disconnected;
+                    };
+
+                    trace!("Sending ingest batch, len: {}", batch.len());
+                    let reply_rx = permit
+                        .send_rpc(IngestRequest::from(Arc::clone(&batch)), Some(self.partition.into()))
+                        .expect("encoding version to match");
+
+                    let batch = IngressBatch{records: batch, trackers, reply_rx: Some(reply_rx)};
+                    self.inflight.push_back(batch);
+                }
+                Some(result) = head => {
+                    match result {
+                        Ok(IngestResponse::Ack) => {
+                            let batch = self.inflight.pop_front().expect("not empty");
+                            batch.committed();
+                        }
+                        Ok(response) => {
+                            // Handle any other error as a connection loss
+                            // and retry all inflight batches.
+                            debug!("Ingest response '{:?}'", response);
+                            break SessionState::Disconnected;
+                        }
+                        Err(_err) => {
+                            // we can assume that for any error
+                            // we need to retry all the inflight bathes.
+                            // special case for load shedding we could
+                            // throttle the stream a little bit then
+                            // speed up over a period of time.
+
+                            break SessionState::Disconnected;
+                        }
+                    }
+                }
+                _ = ctx.cancelled() => {
+                    // relies on auto drain and drop of the inflight
+                    // batches to notify callers that records has been
+                    // cancelled.
+                    return SessionState::Shutdown;
+                }
+            }
+        };
+
+        // don't lose the buffered batch in case of ChunksTimeout
+        if let Either::Left(chunked) = chunked.as_pin_mut() {
+            let remainder = chunked.into_remainder();
+            if !remainder.is_empty() {
+                let (trackers, records): (Vec<_>, Vec<_>) = remainder.into_iter().unzip();
+                let batch: Arc<[IngestRecord]> = Arc::from(records);
+                self.inflight.push_back(IngressBatch {
+                    records: batch,
+                    trackers,
+                    reply_rx: None,
+                });
+            }
+        }
+
+        state
+    }
+}
+
+struct SessionManagerInner<T> {
+    networking: Networking<T>,
+    partition_routing: PartitionRouting,
+    opts: SessionOptions,
+    ctx: CancellationToken,
+    published: ArcSwap<HashMap<PartitionId, SessionHandle>>,
+    locks: DashMap<PartitionId, OnceLock<SessionHandle>>,
```

Comment:
Has the `DashMap` become a contention point which motivated you to introduce the `published` `ArcSwap`? Using a `Mutex<HashMap<PartitionId, SessionHandle>>` wouldn't work because of the expected access patterns?

Reply:
I did not measure performance with using dahsmap only directly to be honest. I was trying to optimise mainly for read-only access and avoid any locks in that pass. Since the write is only needed when starting a new partition session it was the only pass that needs locking. 

Thinking about it now this might have been a premature optimization. i can simplify this by using dashmaps directly and see if there is an impact on performance. Using dashmap also will have the benefit of not cloning the SessionHandle each time.